diff --git a/dist/assets/build/generate.js b/dist/assets/build/generate.js
index 1460e8fe8b62bbed60fdb020b8ea0188e3a3dfa8..1e314a587b5f0eaba522b7b46e568ddab49ea9f2 100644
--- a/dist/assets/build/generate.js
+++ b/dist/assets/build/generate.js
@@ -53,7 +53,10 @@ async function generateImagesForPath(originalFilePath, transformsAndPath, env) {
   for (const [_, transform] of transformsAndPath.transforms) {
     await generateImage(transform.finalPath, transform.transform);
   }
-  if (transformsAndPath.originalSrcPath && !globalThis.astroAsset.referencedImages?.has(transformsAndPath.originalSrcPath)) {
+  const originalOutputIsGenerated = Array.from(
+    transformsAndPath.transforms.values()
+  ).some(({ finalPath }) => finalPath === originalFilePath);
+  if (transformsAndPath.originalSrcPath && !globalThis.astroAsset.referencedImages?.has(transformsAndPath.originalSrcPath) && !originalOutputIsGenerated) {
     try {
       if (transformsAndPath.originalSrcPath) {
         env.logger.debug(
diff --git a/dist/assets/utils/hash.js b/dist/assets/utils/hash.js
index 171f39d58241d3dad4f32ab630cd9c6d4855f3fc..58b55384f8626e7cfd1f4dabbb5643e26ba579cf 100644
--- a/dist/assets/utils/hash.js
+++ b/dist/assets/utils/hash.js
@@ -36,7 +36,7 @@ function propsToFilename(filePath, transform, hash) {
   }
   const prefixDirname = isESMImportedImage(transform.src) ? dirname(filePath) : "";
   let outputExt = transform.format ? `.${transform.format}` : ext;
-  return `${prefixDirname}/${filename}_${hash}${outputExt}`;
+  return `${prefixDirname}/${filename}${outputExt}`;
 }
 function hashTransform(transform, imageService, propertiesToHash) {
   const hashFields = propertiesToHash.reduce(
diff --git a/dist/assets/vite-plugin-assets.js b/dist/assets/vite-plugin-assets.js
index 25b8c3f6037f3c7888eec376545c88a1f2167bb7..ba0b548159cc330aea2b75fd7ade2a4bbdb18437 100644
--- a/dist/assets/vite-plugin-assets.js
+++ b/dist/assets/vite-plugin-assets.js
@@ -1,4 +1,4 @@
-import { extname } from "node:path";
+import { extname, relative } from "node:path";
 import MagicString from "magic-string";
 import picomatch from "picomatch";
 import { AstroError, AstroErrorData } from "../core/errors/index.js";
@@ -277,7 +277,13 @@ function assets({ fs, settings, sync, logger }) {
           if (!assetRegexEnds.test(id)) {
             return;
           }
-          const fileEmitter = shouldEmitFile ? (opts) => emitClientAsset(this, opts) : void 0;
+          const originalFileName = normalizePath(relative(resolvedConfig.root, id));
+          const imagePath = originalFileName.match(/(?:^|\/)assets\/images\/(.+)$/)?.[1];
+          const fileEmitter = shouldEmitFile ? (opts) => emitClientAsset(this, {
+            ...opts,
+            ...imagePath ? { fileName: joinPaths(settings.config.build.assets, "images", imagePath) } : {},
+            originalFileName
+          }) : void 0;
           const imageMetadata = await emitImageMetadata(id, fileEmitter);
           if (!imageMetadata) {
             throw new AstroError({
